File: sdks/go/client_test.go
package swissarmynoife_test

import (
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"

	swissarmynoife "github.com/tycheung/swissarmynoife-sdk"
)

func TestNewClientStripsSlash(t *testing.T) {
	c := swissarmynoife.NewClient("http://127.0.0.1:8787/")
	if c.BaseURL != "http://127.0.0.1:8787" {
		t.Fatalf("BaseURL=%q", c.BaseURL)
	}
}

func TestHealth(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/health" {
			t.Fatalf("path=%s", r.URL.Path)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{"ok": true})
	}))
	defer srv.Close()

	c := swissarmynoife.NewClient(srv.URL)
	out, err := c.Health()
	if err != nil {
		t.Fatal(err)
	}
	m := out.(map[string]any)
	if m["ok"] != true {
		t.Fatalf("%v", out)
	}
}

func TestListHelpers(t *testing.T) {
	cases := []struct {
		path string
		body map[string]any
		call func(*swissarmynoife.Client) (any, error)
	}{
		{"/v1/sak/modules", map[string]any{"modules": []any{}}, (*swissarmynoife.Client).ListModules},
		{"/v1/sak/compute/work", map[string]any{"work": []any{}}, (*swissarmynoife.Client).ListWork},
		{"/v1/sak/compute/nodes", map[string]any{"nodes": []any{}}, (*swissarmynoife.Client).ListNodes},
		{"/v1/sak/capacity", map[string]any{"snapshot": map[string]any{"total_ram_mb": 1.0}}, (*swissarmynoife.Client).Capacity},
	}
	for _, tc := range cases {
		t.Run(tc.path, func(t *testing.T) {
			srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
				if r.URL.Path != tc.path {
					t.Fatalf("got %s", r.URL.Path)
				}
				_ = json.NewEncoder(w).Encode(tc.body)
			}))
			defer srv.Close()
			c := swissarmynoife.NewClient(srv.URL)
			out, err := tc.call(c)
			if err != nil {
				t.Fatal(err)
			}
			if out == nil {
				t.Fatal("nil")
			}
		})
	}
}

func TestEnqueueWork(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/sak/compute/work" || r.Method != http.MethodPost {
			t.Fatalf("%s %s", r.Method, r.URL.Path)
		}
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		if body["action"] != "enqueue" || body["kind"] != "echo" {
			t.Fatalf("%v", body)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{"action": "enqueue", "work": map[string]any{"status": "queued"}})
	}))
	defer srv.Close()

	c := swissarmynoife.NewClient(srv.URL)
	out, err := c.EnqueueWork("echo", map[string]any{"n": 1})
	if err != nil {
		t.Fatal(err)
	}
	m := out.(map[string]any)
	if m["action"] != "enqueue" {
		t.Fatalf("%v", out)
	}
}

func TestRequeueAndClaim(t *testing.T) {
	var lastAction string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		lastAction = body["action"].(string)
		_ = json.NewEncoder(w).Encode(map[string]any{"action": lastAction, "work": map[string]any{"id": "w1"}})
	}))
	defer srv.Close()
	c := swissarmynoife.NewClient(srv.URL)

	if _, err := c.RequeueWork("w1"); err != nil {
		t.Fatal(err)
	}
	if lastAction != "requeue" {
		t.Fatalf("action=%s", lastAction)
	}
	if _, err := c.ClaimWork("n1"); err != nil {
		t.Fatal(err)
	}
	if lastAction != "claim" {
		t.Fatalf("action=%s", lastAction)
	}
	if _, err := c.CompleteWork("w1", "n1", nil); err != nil {
		t.Fatal(err)
	}
	if lastAction != "complete" {
		t.Fatalf("action=%s", lastAction)
	}
	if _, err := c.GetWork("w1"); err != nil {
		t.Fatal(err)
	}
	if lastAction != "get" {
		t.Fatalf("action=%s", lastAction)
	}
}

func TestListWorkFiltered(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		if body["action"] != "list" || body["status"] != "queued" {
			t.Fatalf("%v", body)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{"work": []any{}})
	}))
	defer srv.Close()
	c := swissarmynoife.NewClient(srv.URL)
	out, err := c.ListWorkFiltered(map[string]any{"status": "queued"})
	if err != nil {
		t.Fatal(err)
	}
	m := out.(map[string]any)
	if _, ok := m["work"]; !ok {
		t.Fatalf("%v", out)
	}
}

File: sdks/dart/test/client_test.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:swissarmynoife/swissarmynoife.dart';
import 'package:test/test.dart';

void main() {
  test('baseUrl strips slash', () {
    final c = SakClient('http://127.0.0.1:8787/');
    expect(c.baseUrl, 'http://127.0.0.1:8787');
  });

  test('health', () async {
    final mock = MockClient((req) async {
      expect(req.url.path, '/health');
      return http.Response(jsonEncode({'ok': true}), 200);
    });
    final c = SakClient('http://example.test', httpClient: mock);
    expect((await c.health())['ok'], true);
  });

  test('list helpers', () async {
    final cases = {
      'listModules': ['/v1/sak/modules', {'modules': []}],
      'listWork': ['/v1/sak/compute/work', {'work': []}],
      'listNodes': ['/v1/sak/compute/nodes', {'nodes': []}],
      'capacity': [
        '/v1/sak/capacity',
        {
          'snapshot': {'total_ram_mb': 1}
        }
      ],
    };
    for (final entry in cases.entries) {
      final path = entry.value[0] as String;
      final body = entry.value[1] as Map<String, dynamic>;
      final mock = MockClient((req) async {
        expect(req.url.path, path);
        return http.Response(jsonEncode(body), 200);
      });
      final c = SakClient('http://example.test', httpClient: mock);
      final out = await switch (entry.key) {
        'listModules' => c.listModules(),
        'listWork' => c.listWork(),
        'listNodes' => c.listNodes(),
        _ => c.capacity(),
      };
      expect(out, isA<Map<String, dynamic>>());
    }
  });

  test('enqueue and claim', () async {
    final actions = <String>[];
    final mock = MockClient((req) async {
      final body = jsonDecode(req.body) as Map<String, dynamic>;
      actions.add(body['action'] as String);
      return http.Response(
          jsonEncode({
            'action': 'ok',
            'work': {'id': 'w1'}
          }),
          200);
    });
    final c = SakClient('http://example.test', httpClient: mock);
    await c.enqueueWork('echo', {'n': 1});
    await c.requeueWork('w1');
    await c.claimWork('n1');
    await c.completeWork('w1', 'n1');
    await c.getWork('w1');
    expect(actions, ['enqueue', 'requeue', 'claim', 'complete', 'get']);
  });
}
